### PR TITLE
Adding "unsafe" option for reading dted

### PR DESCRIPTION
**Pull request by amclauth**
Many dted0 files I've come across have bad or empty fields in various headers (especially edition and dates) and will not be able to be read with the current implementation. Also, many have void data, which throws a warning, but if you know the file and don't want to see the warning, this adds a way to handle that as well.

### CODE REVIEW

**Comment by bbonenfant** [comment]
Hello, thank you for your interest in this package. You raise a valid criticism that, in its current state, this package will only parse valid DTED files according to [this specification](https://www.dlr.de/eoc/Portaldata/60/Resources/dokumente/7_sat_miss/SRTM-XSAR-DEM-DTED-1.1.pdf). A lot of this data isn't actually needed when extracting the elevation data.

However, some of the fields you are proposing to make optional are pertinent to parsing the data, such as the `DataSetIdentification.shape` field. I would be hesitant to make these fields default-able -- especially without a test. I presume you wouldn't be able to share any of your DTED0 data as test data.

How exactly are you intending to use this package on your DTED data? If you already have access to the pertinent metadata outside this file and only need access to the grid of parsed elevation data, then we could probably implement a pass-through to get this data. Something like `dted.extract_grid(file, shape=(N,M))`

**Comment by amclauth** [comment]
Valid criticisms ... I thought of similar things but didn't know what your
preferences would be so I shotgunned all of them as a first attempt. You're
right that I can't pull the dted out of the system it's on so it'd be hard
to help build valid tests. I can't even move the code between the systems
and here :).

What I can pretty easily do through is test ideas (code change locally)
against the data I have to see if it'd work for the data in *our* system. I
couldn't begin to say whether that's representative. But for example, shape
never seems to not be set, so I could revert the ValueError tests on shape
in uhl and dsi (did that locally) with no problems at all while still
getting all the other fields. In my own files we don't have access to the
metadata in advance, so it's now double-obvious that for me, shape was not
a problem.

I liked the idea of "get as many of the header fields as you can from the
file", so I just put checks on *all* the fields that could fail due to
casting. Totally respect the "except fields I need to parse the file" idea,
too. I think that'd be a nice compromise ... the "minimally viable"
self-parsing dted file ... that would help keep the flavor of the package
the same as well.

I could probably do something stupid and simple to characterize these "bad"
dted files ... add a print statement in each valueerror check that says,
for example, "dsi:shape", and then sort those to see which fields
*actually* tend
to be nullable? But that'd just be for the datasets we have on hand. "MVSP"
would be a more rigorous test (and the files I have obviously "work" to
self-parse, so it'd be valid for my use-case as well).

And thanks for creating this package! It's a solid piece of work. I love
seeing other people's answers to these kinds of problems. I'm trying to
move teams to the idea of using more public repos like this where there's a
chance people from all over will help generate better and better solutions!

**Comment by bbonenfant** [comment]
OK, this gives me a much better picture of your use-case. I don't think there's anything "unsafe" about what you are asking for so its possible we can have most of these fields be nullable by default and require only what's needed. Let me spend some time thinking about this tonight and see what I come up with

**Comment by amclauth** [comment]
Ran a quick test to see where the problem fields were (again,
representative only of my own data, and I like the idea of a "minimally
parseable" option that doesn't throw warnings):

25594 files

2 times:
uhl: vertical_accuracy

12862 times:
acc: absolute_horizontal
acc: absolute_vertical
acc: relative_horizontal
acc: relative_vertical
dsi: compilation_date
dsi: edition
dsi: maintenance_date
dsi: merge_date
dsi: specification_date

I didn't run into any void statements on this dataset (unless I mistyped
somehow), but saw those somewhere and would want to suppress them without
having to override warning as being ocean altitudes from what I can see
(and you note the same).

**Comment by amclauth** [comment]
Just noticed ... the warns show up in *some *dted1 and dted2 files, never
in my dted0 files.

And FYI, just for fun of sharing, the two things I've wrapped with this are
to:

1) Pass in a list of lat/lon and get back altitudes (in the original order)
where the lat/lon list gets sorted to dted tile block so I can load in a
single file at a time and hit it for every altitude I want and then throw
it away

2) Use the RegularGridInterpolator from scipy to do interpolation across
the tile. I'm not yet sure if it's meaningfully different to try and use
blocks smaller than a tile and I may not care if this is "fast enough" to
interpolate altitude for any lat/lon point. For that, I really only need
the tile.data (transposed), but the shape and origin lat/lon are helpful
shorthands to get into it.

**Comment by bbonenfant** [comment]
I made some changes to your pull request (I hope that is OK). Based on the problem fields in your dataset, I made all the type-casted fields optional by default except for the fields `uhl.shape`, `dsi.shape`, `dsi.latitude_interval`, and `dsi.longitude_interval`. Please try these changes on your dataset to verify everything still works.

The VoidDataWarning is thrown using the stdlib `warnings` module. You should be able to suppress the warning with:
```python
import warnings
from dted.errors import VoidDataWarning
warnings.simplefilter(action="ignore", category=VoidDataWarning)
```
I actually thought this code snippet was included within the README but I must have forgotten it. I'll make sure to add it in. I'm also intrigued that your DTED0 data doesn't emit the warning. If you wouldn't mind doing some investigation, I'd be curious to know if your DTED0 files have valid data for these "void" regions or if they are using a different null value.

**Comment by amclauth** [comment]
Of course it's 100% ok to change my commit .. it's your project! :)

This change should very much work for my dataset but I'll try to give it a
shot today and confirm

Stylistically I do like the try_int and try_float a lot better ... far
cleaner and just makes the "minimally parseable" object implicit instead of
explicit, and InvalidFileError is a better thing to catch than ValueError
(which could mean *so *much).

The warning, I personally don't love because to ignore it you have to
ignore this kind of warning for the whole project. Granted, this is a
custom warning, but it's "VoidData" which doesn't *sound* custom as
opposed to something like "VoidDtedAltitudeWarning" ... more obvious when
you see at the top of a project. Personally, I'd want a way to skip those
voids within the area of code calling Tile, whether that's an optional flag
or if instead of throwing a void it throws an error that's catchable right
there.

BUT ... this is 100% your project and your style and you *do* already have
this as a workaround so there aren't warnings popping up if I use this for
a large dataset that has these fields set to void. I'm just expressing my
opinion there :)

My dted0 appears to have 0's anywhere the data isn't valid and definitely
for the areas that dted1 and dted2 have voids. I'm not sure why in this
dataset *some* files have voids and others don't. Just what we got. Our c++
and java versions of dted readers just ignored them ... apparently the
assumption in those code blocks is that they're all water / invalid
altitudes.

**Comment by amclauth** [comment]
I put these changes in and yes, assuming I typed everything correctly, I
can confirm that these would work nicely!

File: dted/records/dsi.py [hunk 1]
```diff
@@ -151,12 +152,32 @@ def from_bytes(cls, data: bytes) -> "DataSetIdentification":
...
+
+        shape = (try_int(buffered_data.read(4)), try_int(buffered_data.read(4)))[::-1]
+        if shape[0] is None or shape[1] is None:
+            raise InvalidFileError(
+                "The shape of the gridded data must be specified in the "
+                "DataSetIdentification section of the DTED file. "
+            )
+
+        coverage = try_float(buffered_data.read(2))
+        if coverage is not None:
```
Comment (amclauth) [inline dted/records/dsi.py]: The if doesn't look like it's needed here. If try_float returns None, the next line, "coverage=1 if coverage == 0 else coverage" would still keep coverage at None

Comment (bbonenfant) [reply dted/records/dsi.py]: good catch!

**Comment by bbonenfant** [comment]
After reviewing the warnings module, I'm not very impressed by it's functionality for suppressing warning messages for specific blocks of code. Therefore I've added `warn` kwargs to `Tile.__init__` and `Tile.load_data` to allow suppressing the VoidDataWarning. If that's satisfactory to you then I'm happy to merge this in and publish v1.0.4.

**Comment by amclauth** [comment]
Makes sense to me. Good job!

**Pull request merged**